modules/frontend/querysharding.go

```diff
@@ -68,29 +72,12 @@ func (s shardQuery) Do(r *http.Request) (*http.Response, error) {
 		reqs[i].RequestURI = querierPrefix + reqs[i].URL.RequestURI() + queryDelimiter + q.Encode()
 	}
 
-	rrs, err := DoRequests(r.Context(), s.next, reqs)
+	rrs, err := doRequests(r.Context(), s.next, reqs)
 	if err != nil {
 		return nil, err
 	}
 
-	// todo: add merging logic here if there are more than one results
-	var errCode int
-	var errBody io.ReadCloser
-	for _, rr := range rrs {
-		if rr.Response.StatusCode == http.StatusOK {
-			return rr.Response, nil
-		}
-		if rr.Response.StatusCode > errCode {
-			errCode = rr.Response.StatusCode
-			errBody = rr.Response.Body
-		}
-	}
-
-	return &http.Response{
-		StatusCode: errCode,
-		Body:       errBody,
-		Header:     http.Header{},
-	}, nil
+	return mergeResponses(rrs)
```

Comment:
If we decide that 5xx always takes precedence, it might be clearer to split that into a separate func like `checkForErrors(rrs)`.

Reply:
The function also combines traces that might be split across blocks, so `mergeResponses` seems more appropriate?